### PR TITLE
Sim3 pose comparison

### PR DESCRIPTION
**Pull request by johnwlambert**
- Make the target poses the first arg in pose alignment. Frame "A" is considered the first class frame (ground truth frame), and we move the inputs from frame "B" into frame "A" with `aSb`
- Fix unit tests from previous DA PR
- Remove old notation where there were two "world frames", because there can only be one such frame

### CODE REVIEW

**Review by akshay-krishnan — APPROVED** [review]
LGTM!

File: README.md [hunk 1]
```diff
@@ -60,7 +60,7 @@ GTSFM is designed in an extremely modular way. Each module can be swapped out wi
         - `translation`: translation averaging implementations (1d-SFM, etc)
     - `bundle`: bundle adjustment implementations
     - `common`: basic classes used through GTSFM, such as `Keypoints`, `Image`, `SfmTrack2d`, etc
-    - `data_association`: 
+    - `data_association`: 3d point triangulation or w/o RANSAC, from 2d point-tracks
```
Comment (ayushbaid) [inline README.md:63]: "with optional RANSAC"? or  "w/ or w/o RANSAC"?

File: gtsfm/utils/geometry_comparisons.py [hunk 2]
```diff
@@ -30,40 +30,40 @@ def align_rotations(input_list: List[Rot3], ref_list: List[Rot3]) -> List[Rot3]:
...
+    We assume the two trajectories are of the exact same length.
 
+    Args:
+        aTi_list: reference poses in frame "a" which are the targets for alignment
+        bTi_list: input poses which need to be aligned to frame "a"
+            
     Returns:
-        transformed poses which have the same origin and scale as reference (now living in world-2 frame)
+        aTi_list_: transformed input poses previously "bTi_list" but now which
+            have the same origin and scale as reference (now living in "a" frame)
```
Comment (ayushbaid) [inline gtsfm/utils/geometry_comparisons.py:45]: I think this convention is opposite to what we have in the other functions. The first input is transformed to reside in the frame of second input for all other functions.

File: gtsfm/utils/geometry_comparisons.py [hunk 3]
```diff
@@ -30,40 +30,40 @@ def align_rotations(input_list: List[Rot3], ref_list: List[Rot3]) -> List[Rot3]:
...
     """
-    # match the scales first
-    wTi0 = input_list[0]
-    input_distances = np.array([np.linalg.norm((wTi.between(wTi0)).translation()) for wTi in input_list[1:]])
+    n_to_align = len(aTi_list)
+    assert len(aTi_list) == len(bTi_list)
+    assert n_to_align >= 2, "SIM(3) alignment uses at least 2 frames"
```
Comment (ayushbaid) [inline gtsfm/utils/geometry_comparisons.py:49]: Do you think raising `ValueError` is more appropriate than raising `AssertionError`

File: gtsfm/utils/geometry_comparisons.py [hunk 4]
```diff
@@ -30,40 +30,40 @@ def align_rotations(input_list: List[Rot3], ref_list: List[Rot3]) -> List[Rot3]:
...
-    # now match origin
-    w1Ti0 = scaled_list[0]
-    i0Tw1 = w1Ti0.inverse()
-    w2Ti0_ = ref_list[0]
-    # origin transform -- map the origin of the input list to the reference list
-    w2Tw1 = w2Ti0_.compose(i0Tw1)
+        aTi_list_.append(aSb.transformFrom(bTi))
 
-    scaled_shifted_list = [w2Tw1.compose(w1Ti) for w1Ti in scaled_list]
+    print("Trajectory alignment complete.")
```
Comment (ayushbaid) [inline gtsfm/utils/geometry_comparisons.py:64]: logging instead of print? Or just remove it altogether?

**Review by ayushbaid — APPROVED** [review]
Look good except one suggestion on the ordering of the input. 

I think you also fixed the tests which are failing in master, right? These tests were failing due to the DA metrics PR.

One more thing I'd like to see is how the metrics change with this new alignment.

File: gtsfm/utils/geometry_comparisons.py (hunk 2 above)
Comment (johnwlambert) [reply gtsfm/utils/geometry_comparisons.py:45]: thanks for the review @ayushbaid . i actually did that intentionally, and i think i updated all the API references to update them.

The main idea is that if we have two frames A and B, frame "A" seems like the first class citizen. So we would estimate aSb, to force frame "B" elements to move into frame "A".
```python
aTi = aSb * bTi
```
So I made the args (aTi_list, bTi_list)

Comment (ayushbaid) [reply gtsfm/utils/geometry_comparisons.py:45]: Ok got it. I meant other functions like `align_rotations`.

We should follow the same convention for every function.

File: gtsfm/utils/geometry_comparisons.py (hunk 4 above)
Comment (johnwlambert) [reply gtsfm/utils/geometry_comparisons.py:64]: good point, i'll fix this

File: gtsfm/utils/geometry_comparisons.py (hunk 2 above)
Comment (johnwlambert) [reply gtsfm/utils/geometry_comparisons.py:54]: Should be updated now for `align_rotations()` as well

File: gtsfm/utils/geometry_comparisons.py (hunk 3 above)
Comment (johnwlambert) [reply gtsfm/utils/geometry_comparisons.py:58]: Not too sure to be honest

**Comment by ayushbaid** [comment]
Looks good now. Feel free to merge once we have numbers from a dataset to show improvements in error, if any.

**Comment by johnwlambert** [comment]
Thanks @ayushbaid . Here are the stats when I run:
```python
python gtsfm/runner/run_scene_optimizer_argoverse.py --dataset_dir ~/Downloads/visual-odometry-tutorial/full_log --stride 5 --max_num_imgs 10
[2021-03-16 15:51:40,841 INFO run_scene_optimizer_argoverse.py line 95 63314] Namespace(camera_name='ring_front_center', dataset_dir='/Users/johnlambert/Downloads/visual-odometry-tutorial/full_log', log_id='273c1883-673a-36bf-b124-88311b1a80be', max_lookahead_sec=2, max_num_imgs=10, stride=5)
```

Using old alignment:
```python
    "rotation_averaging_angle_deg": {
        "median_error": 0.06949349547825891,
        "min_error": 0.0330567736410538,
        "max_error": 0.12980091454465748,
}
    "translation_averaging_distance": {
        "median_error": 0.22903253235700852,
        "min_error": 0.054435723335768356,
        "max_error": 0.546566357877385,
    },
    "translation_to_direction_angle_deg": {
        "median_error": 0.3375493876771237,
        "min_error": 0.007295816082640759,
        "max_error": 1.0173310275651082,
```
Using new Sim(3) alignment:
```python
{
    "rotation_averaging_angle_deg": {
        "median_error": 0.07248240344151745,
        "min_error": 0.030406933538419115,
        "max_error": 0.11827609453611045,

    "translation_averaging_distance": {
        "median_error": 0.4906749323356928,
        "min_error": 0.07231437604763874,
        "max_error": 1.6007145553706401,

    "translation_to_direction_angle_deg": {
        "median_error": 0.2716479893914394,
        "min_error": 0.01655628350688935,
        "max_error": 1.3441434828326693,
```

**Comment by johnwlambert** [comment]
The angular translation errors go down, but distance errors go up. rotation averaging distances seem about the same

**Comment by johnwlambert** [comment]
This was a straight sequence. Let me check a curved sequence

**Pull request merged**